# Pre-Processing and Training Data Development

Here we have 3 simple goals: <br>

1) Create dummy features for categorical data necessary <br>
2) Scale numerical variables if necessary <br>
3) Split data into training and test sets for modelling <br>

In [1]:
# Import necessary packages:

import pandas as pd
import numpy as np


In [2]:
# First let's load in our data

df = pd.read_csv(r'C:\Users\bronc\Downloads\Capstone 2\Real_Estate_Sales_2001-2018(clean).csv')
df.head()

Unnamed: 0  Unnamed: 0.1  SerialNumber  ListYear DateRecorded        Town  \
0           0             0        110540      2011   2012-04-03    Stamford   
1           1             1        120025      2012   2012-10-05   Greenwich   
2           2             3         60173      2006   2006-12-28     Windsor   
3           3             5         14539      2014   2015-05-28     Milford   
4           4             9        160629      2016   2017-01-31  Bridgeport   

                          Address  AssessedValue  SaleAmount  SalesRatio  \
0             56 CHERRY HILL ROAD         795870      690000    0.866976   
1  78 BALDWIN FARMS SOUTH, GREENW        1925560     3224000    1.674318   
2                     7 ALFORD DR         189630      309000    1.629489   
3               170 MEADOWSIDE RD         147340      150000    1.018053   
4                  75 EDGEMOOR RD         163380      180000    1.101726   

  PropertyType ResidentialType  Month  
0  Residential   Single Family      4  
1  Residential   Single Family     10  
2  Residential   Single Family     12  
3  Residential   Single Family      5  
4  Residential   Single Family      1

### Create dummy features

In [3]:
#Looking through the data I think it makes the most sense to create dummy features for Town and ResidentialType
# However with towns since there are so many I want to aggregate those with less than 1000 observations into an other column
# To do this we must first get the value counts for Town
df['Town'].value_counts()

Bridgeport       20271
Waterbury        16554
Stamford         15131
New Haven        13044
West Hartford    12669
                 ...  
Franklin           264
Eastford           249
Hartland           240
Canaan             217
Union              143
Name: Town, Length: 169, dtype: int64

In [4]:
df['Town'] = df['Town'].replace('Franklin', 'OtherTown')
df['Town'] = df['Town'].replace('Eastford', 'OtherTown')
df['Town'] = df['Town'].replace('Hartland', 'OtherTown')
df['Town'] = df['Town'].replace('Canaan', 'OtherTown')
df['Town'] = df['Town'].replace('Union', 'OtherTown')
df['Town'] = df['Town'].replace('Bozrah', 'OtherTown')
df['Town'] = df['Town'].replace('Colebrook', 'OtherTown')
df['Town'] = df['Town'].replace('Cornwall', 'OtherTown')
df['Town'] = df['Town'].replace('Scotland', 'OtherTown')
df['Town'] = df['Town'].replace('Union', 'OtherTown')
df['Town'] = df['Town'].replace('Norfolk', 'OtherTown')
df['Town'] = df['Town'].replace('Chaplin', 'OtherTown')
df['Town'] = df['Town'].replace('Hampton', 'OtherTown')
df['Town'] = df['Town'].replace('Warren', 'OtherTown')
df['Town'] = df['Town'].replace('Bridgewater', 'OtherTown')
df['Town'] = df['Town'].replace('North Canaan', 'OtherTown')
df['Town'] = df['Town'].replace('Voluntown', 'OtherTown')
df['Town'] = df['Town'].replace('Morris', 'OtherTown')
df['Town'] = df['Town'].replace('Sprague', 'OtherTown')
df['Town'] = df['Town'].replace('Roxbury', 'OtherTown')
df['Town'] = df['Town'].replace('Kent', 'OtherTown')
df['Town'] = df['Town'].replace('Barkhamsted', 'OtherTown')
df['Town'] = df['Town'].replace('Andover', 'OtherTown')
df['Town'] = df['Town'].replace('Bethlehem', 'OtherTown')
df['Town'] = df['Town'].replace('Lyme', 'OtherTown')
df['Town'] = df['Town'].replace('Chester', 'OtherTown')
df['Town'] = df['Town'].replace('Ashford', 'OtherTown')
df['Town'] = df['Town'].replace('Pomfret', 'OtherTown')
df['Town'] = df['Town'].replace('Lisbon', 'OtherTown')
df['Town'] = df['Town'].replace('Middlefield', 'OtherTown')
df['Town'] = df['Town'].replace('Salem', 'OtherTown')
df['Town'] = df['Town'].replace('Deep River', 'OtherTown')
df['Town'] = df['Town'].replace('Willington', 'OtherTown')
df['Town'] = df['Town'].replace('Sharon', 'OtherTown')
df['Town'] = df['Town'].replace('Sterling', 'OtherTown')
df['Town'] = df['Town'].replace('Harwinton', 'OtherTown')
df['Town'] = df['Town'].replace('Preston', 'OtherTown')
df['Town'] = df['Town'].replace('Sherman', 'OtherTown')
df['Town'] = df['Town'].replace('Canterbury', 'OtherTown')
df['Town'] = df['Town'].replace('Goshen', 'OtherTown')
df['Town'] = df['Town'].replace('Beacon Falls', 'OtherTown')
df['Town'] = df['Town'].replace('Washington', 'OtherTown')
df['Town'] = df['Town'].replace('Salisbury', 'OtherTown')
df['Town'] = df['Town'].replace('Bolton', 'OtherTown')
df['Town'] = df['Town'].replace('Bethany', 'OtherTown')
df['Town'] = df['Town'].replace('North Stonington', 'OtherTown')

In [5]:
df['Town'].value_counts()

OtherTown       25864
Bridgeport      20271
Waterbury       16554
Stamford        15131
New Haven       13044
                ...  
Middlebury       1161
Marlborough      1115
New Hartford     1076
East Granby      1040
Columbia         1024
Name: Town, Length: 125, dtype: int64

In [6]:
dummy_town = pd.get_dummies(df['Town'])
dummy_res = pd.get_dummies(df['ResidentialType'])

In [7]:
dummy_town.head()

Ansonia  Avon  Berlin  Bethel  Bloomfield  Branford  Bridgeport  Bristol  \
0        0     0       0       0           0         0           0        0   
1        0     0       0       0           0         0           0        0   
2        0     0       0       0           0         0           0        0   
3        0     0       0       0           0         0           0        0   
4        0     0       0       0           0         0           1        0   

   Brookfield  Brooklyn  ...  Wethersfield  Wilton  Winchester  Windham  \
0           0         0  ...             0       0           0        0   
1           0         0  ...             0       0           0        0   
2           0         0  ...             0       0           0        0   
3           0         0  ...             0       0           0        0   
4           0         0  ...             0       0           0        0   

   Windsor  Windsor Locks  Wolcott  Woodbridge  Woodbury  Woodstock  
0        0              0        0           0         0          0  
1        0              0        0           0         0          0  
2        1              0        0           0         0          0  
3        0              0        0           0         0          0  
4        0              0        0           0         0          0  

[5 rows x 125 columns]

In [8]:
dummy_res.sample(5)

Four Family  Multi Family  Single Family  Three Family  Two Family
405346            0             0              1             0           0
516079            0             0              1             0           0
166090            0             0              1             0           0
455166            0             0              1             0           0
423672            0             0              1             0           0

In [9]:
# With those made and verified to have worked let's put them back with our main dataframe

df = pd.concat([df, dummy_town], axis = 1)
df.head()

Unnamed: 0  Unnamed: 0.1  SerialNumber  ListYear DateRecorded        Town  \
0           0             0        110540      2011   2012-04-03    Stamford   
1           1             1        120025      2012   2012-10-05   Greenwich   
2           2             3         60173      2006   2006-12-28     Windsor   
3           3             5         14539      2014   2015-05-28     Milford   
4           4             9        160629      2016   2017-01-31  Bridgeport   

                          Address  AssessedValue  SaleAmount  SalesRatio  ...  \
0             56 CHERRY HILL ROAD         795870      690000    0.866976  ...   
1  78 BALDWIN FARMS SOUTH, GREENW        1925560     3224000    1.674318  ...   
2                     7 ALFORD DR         189630      309000    1.629489  ...   
3               170 MEADOWSIDE RD         147340      150000    1.018053  ...   
4                  75 EDGEMOOR RD         163380      180000    1.101726  ...   

  Wethersfield Wilton  Winchester  Windham  Windsor  Windsor Locks  Wolcott  \
0            0      0           0        0        0              0        0   
1            0      0           0        0        0              0        0   
2            0      0           0        0        1              0        0   
3            0      0           0        0        0              0        0   
4            0      0           0        0        0              0        0   

   Woodbridge  Woodbury  Woodstock  
0           0         0          0  
1           0         0          0  
2           0         0          0  
3           0         0          0  
4           0         0          0  

[5 rows x 138 columns]

In [10]:
df = pd.concat([df, dummy_res], axis = 1)
df.head()

Unnamed: 0  Unnamed: 0.1  SerialNumber  ListYear DateRecorded        Town  \
0           0             0        110540      2011   2012-04-03    Stamford   
1           1             1        120025      2012   2012-10-05   Greenwich   
2           2             3         60173      2006   2006-12-28     Windsor   
3           3             5         14539      2014   2015-05-28     Milford   
4           4             9        160629      2016   2017-01-31  Bridgeport   

                          Address  AssessedValue  SaleAmount  SalesRatio  ...  \
0             56 CHERRY HILL ROAD         795870      690000    0.866976  ...   
1  78 BALDWIN FARMS SOUTH, GREENW        1925560     3224000    1.674318  ...   
2                     7 ALFORD DR         189630      309000    1.629489  ...   
3               170 MEADOWSIDE RD         147340      150000    1.018053  ...   
4                  75 EDGEMOOR RD         163380      180000    1.101726  ...   

  Windsor Locks Wolcott  Woodbridge  Woodbury  Woodstock  Four Family  \
0             0       0           0         0          0            0   
1             0       0           0         0          0            0   
2             0       0           0         0          0            0   
3             0       0           0         0          0            0   
4             0       0           0         0          0            0   

   Multi Family  Single Family  Three Family  Two Family  
0             0              1             0           0  
1             0              1             0           0  
2             0              1             0           0  
3             0              1             0           0  
4             0              1             0           0  

[5 rows x 143 columns]

I ran these in that order to verify that the columns had been successfully added

### Scale numerical variables

This will be implemented in the modeling stage

### Split data into training and test sets for modelling

In [11]:
#Import the necessary package
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

We'll be seeing how our data does at predicting SaleAmount as that has the most interesting real world application

In [12]:
X = df.iloc[:, 7:144]
y = df[['SaleAmount']]

In [13]:
X.drop(columns = 'SaleAmount', inplace = True)
X.head()

AssessedValue  SalesRatio PropertyType ResidentialType  Month  Ansonia  \
0         795870    0.866976  Residential   Single Family      4        0   
1        1925560    1.674318  Residential   Single Family     10        0   
2         189630    1.629489  Residential   Single Family     12        0   
3         147340    1.018053  Residential   Single Family      5        0   
4         163380    1.101726  Residential   Single Family      1        0   

   Avon  Berlin  Bethel  Bloomfield  ...  Windsor Locks  Wolcott  Woodbridge  \
0     0       0       0           0  ...              0        0           0   
1     0       0       0           0  ...              0        0           0   
2     0       0       0           0  ...              0        0           0   
3     0       0       0           0  ...              0        0           0   
4     0       0       0           0  ...              0        0           0   

   Woodbury  Woodstock  Four Family  Multi Family  Single Family  \
0         0          0            0             0              1   
1         0          0            0             0              1   
2         0          0            0             0              1   
3         0          0            0             0              1   
4         0          0            0             0              1   

   Three Family  Two Family  
0             0           0  
1             0           0  
2             0           0  
3             0           0  
4             0           0  

[5 rows x 135 columns]

In [14]:
y.head()

SaleAmount
0      690000
1     3224000
2      309000
3      150000
4      180000

In [15]:
X.shape

(574715, 135)

In [16]:
y.shape

(574715, 1)

Perform train test split. I decided to test it out with a value of 30% for the test split

In [17]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 123)

Verify that the split worked

In [18]:
X_train.shape

(402300, 135)

In [19]:
X_test.shape

(172415, 135)

The split looks good! Next we'll start modelling our data

# Modeling

In [20]:
# Import necessary packages

from sklearn import linear_model
from sklearn.metrics import mean_squared_error, r2_score
import statsmodels.api as sm
from sklearn.preprocessing import StandardScaler

After looking at our data and some of the modeling techniques I know I decided to change up the data we'll be inputing

## Model 1

In [21]:
X = df[['AssessedValue', 'ResidentialType']]
X.head()

AssessedValue ResidentialType
0         795870   Single Family
1        1925560   Single Family
2         189630   Single Family
3         147340   Single Family
4         163380   Single Family

In [22]:
X['ResidentialType'].value_counts()

Single Family    524082
Two Family        32932
Three Family      16728
Four Family         904
Multi Family         69
Name: ResidentialType, dtype: int64

In [23]:
X['ResidentialType'] = X['ResidentialType'].replace('Two Family', 'Multi Family')
X['ResidentialType'] = X['ResidentialType'].replace('Three Family', 'Multi Family')
X['ResidentialType'] = X['ResidentialType'].replace('Four Family', 'Multi Family')
X['ResidentialType'].value_counts()

C:\Users\bronc\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
C:\Users\bronc\anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\Users\bronc\anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in

Single Family    524082
Multi Family      50633
Name: ResidentialType, dtype: int64

In [24]:
X = pd.get_dummies(X)
X.head()

AssessedValue  ResidentialType_Multi Family  ResidentialType_Single Family
0         795870                             0                              1
1        1925560                             0                              1
2         189630                             0                              1
3         147340                             0                              1
4         163380                             0                              1

In [25]:
scaler = StandardScaler()

In [26]:
scaler.fit(X)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [27]:
X_scaled = scaler.transform(X)

In [28]:
X_scaled = sm.add_constant(X_scaled)

Re-do Train Test split

In [29]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size = 0.3, random_state = 123)

Verify that it worked again

In [30]:
X_train.shape

(402300, 4)

In [31]:
X_test.shape

(172415, 4)

Create and fit the model

In [32]:
Model1 = sm.OLS(y_train.astype(float), X_train.astype(float))
Model1_results = Model1.fit()

Get the results

In [33]:
Model1_results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             SaleAmount   R-squared:                       0.736
Model:                            OLS   Adj. R-squared:                  0.736
Method:                 Least Squares   F-statistic:                 5.620e+05
Date:                Wed, 24 Mar 2021   Prob (F-statistic):               0.00
Time:                        18:28:59   Log-Likelihood:            -5.4901e+06
No. Observations:              402300   AIC:                         1.098e+07
Df Residuals:                  402297   BIC:                         1.098e+07
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       3.356e+05    322.291   1041.397      0.000    3.35e+05    3.36e+05
x1          3.421e+05    325.024   1052.656      0.000    3.42e+05    3.43e+05
x2         -3075.9693    162.286    -18.954      0.000   -3394.044   -2757.895
x3          3075.9693    162.286     18.954      0.000    2757.895    3394.044
==============================================================================
Omnibus:                   249295.541   Durbin-Watson:                   1.994
Prob(Omnibus):                  0.000   Jarque-Bera (JB):        272688204.798
Skew:                          -1.521   Prob(JB):                         0.00
Kurtosis:                     130.509   Cond. No.                     2.79e+15
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 1.04e-25. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

Finally let's use Root Mean Squared Error (RMSE) to measure how closely our model fits the data

In [34]:
y_pred = Model1_results.predict(X_test)

In [35]:
def rmse(predictions, targets):
    return np.sqrt(((predictions - targets) ** 2).mean())

In [36]:
matches = pd.DataFrame(y_test)
matches.rename(columns = {'SaleAmount':'actual'}, inplace=True)
matches["predicted"] = y_pred

rmse(matches["actual"], matches["predicted"])

208030.65417067677

Here the R-squared is 0.736 and the root mean squared error seems pretty high. This model is decent but I'm sure we can do better. Next let's try using more variables

## Model 2

In [37]:
X = df.iloc[:, 7:144]
X.head()

AssessedValue  SaleAmount  SalesRatio PropertyType ResidentialType  Month  \
0         795870      690000    0.866976  Residential   Single Family      4   
1        1925560     3224000    1.674318  Residential   Single Family     10   
2         189630      309000    1.629489  Residential   Single Family     12   
3         147340      150000    1.018053  Residential   Single Family      5   
4         163380      180000    1.101726  Residential   Single Family      1   

   Ansonia  Avon  Berlin  Bethel  ...  Windsor Locks  Wolcott  Woodbridge  \
0        0     0       0       0  ...              0        0           0   
1        0     0       0       0  ...              0        0           0   
2        0     0       0       0  ...              0        0           0   
3        0     0       0       0  ...              0        0           0   
4        0     0       0       0  ...              0        0           0   

   Woodbury  Woodstock  Four Family  Multi Family  Single Family  \
0         0          0            0             0              1   
1         0          0            0             0              1   
2         0          0            0             0              1   
3         0          0            0             0              1   
4         0          0            0             0              1   

   Three Family  Two Family  
0             0           0  
1             0           0  
2             0           0  
3             0           0  
4             0           0  

[5 rows x 136 columns]

In [38]:
X.drop(columns = 'SaleAmount', inplace = True)
X.drop(columns = 'PropertyType', inplace = True)
X.drop(columns = 'ResidentialType', inplace = True)
X.head()

AssessedValue  SalesRatio  Month  Ansonia  Avon  Berlin  Bethel  \
0         795870    0.866976      4        0     0       0       0   
1        1925560    1.674318     10        0     0       0       0   
2         189630    1.629489     12        0     0       0       0   
3         147340    1.018053      5        0     0       0       0   
4         163380    1.101726      1        0     0       0       0   

   Bloomfield  Branford  Bridgeport  ...  Windsor Locks  Wolcott  Woodbridge  \
0           0         0           0  ...              0        0           0   
1           0         0           0  ...              0        0           0   
2           0         0           0  ...              0        0           0   
3           0         0           0  ...              0        0           0   
4           0         0           1  ...              0        0           0   

   Woodbury  Woodstock  Four Family  Multi Family  Single Family  \
0         0          0            0             0              1   
1         0          0            0             0              1   
2         0          0            0             0              1   
3         0          0            0             0              1   
4         0          0            0             0              1   

   Three Family  Two Family  
0             0           0  
1             0           0  
2             0           0  
3             0           0  
4             0           0  

[5 rows x 133 columns]

In [39]:
X.describe()

AssessedValue    SalesRatio          Month        Ansonia  \
count   5.747150e+05  5.747150e+05  574715.000000  574715.000000   
mean    2.097660e+05  2.674725e+00       6.794531       0.004971   
std     2.578989e+05  2.323596e+02       3.233131       0.070331   
min     1.000000e+00  8.153814e-07       1.000000       0.000000   
25%     9.505000e+04  1.266434e+00       4.000000       0.000000   
50%     1.425900e+05  1.594454e+00       7.000000       0.000000   
75%     2.240700e+05  2.041974e+00      10.000000       0.000000   
max     4.948370e+06  1.416667e+05      12.000000       1.000000   

                Avon         Berlin         Bethel     Bloomfield  \
count  574715.000000  574715.000000  574715.000000  574715.000000   
mean        0.006984       0.005211       0.005129       0.006339   
std         0.083280       0.072001       0.071437       0.079364   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

            Branford     Bridgeport  ...  Windsor Locks        Wolcott  \
count  574715.000000  574715.000000  ...  574715.000000  574715.000000   
mean        0.005883       0.035271  ...       0.004211       0.004762   
std         0.076474       0.184465  ...       0.064754       0.068845   
min         0.000000       0.000000  ...       0.000000       0.000000   
25%         0.000000       0.000000  ...       0.000000       0.000000   
50%         0.000000       0.000000  ...       0.000000       0.000000   
75%         0.000000       0.000000  ...       0.000000       0.000000   
max         1.000000       1.000000  ...       1.000000       1.000000   

          Woodbridge       Woodbury      Woodstock    Four Family  \
count  574715.000000  574715.000000  574715.000000  574715.000000   
mean        0.002939       0.002944       0.002814       0.001573   
std         0.054131       0.054179       0.052968       0.039629   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

        Multi Family  Single Family   Three Family     Two Family  
count  574715.000000  574715.000000  574715.000000  574715.000000  
mean        0.000120       0.911899       0.029107       0.057301  
std         0.010957       0.283442       0.168105       0.232418  
min         0.000000       0.000000       0.000000       0.000000  
25%         0.000000       1.000000       0.000000       0.000000  
50%         0.000000       1.000000       0.000000       0.000000  
75%         0.000000       1.000000       0.000000       0.000000  
max         1.000000       1.000000       1.000000       1.000000  

[8 rows x 133 columns]

Since we're using month let's make it a dummy variable to allow for better modeling. I'll divide the years into quarters

In [40]:
X['Month'] = X['Month'].replace(1, 'Q1')
X['Month'] = X['Month'].replace(2, 'Q1')
X['Month'] = X['Month'].replace(3, 'Q1')
X['Month'] = X['Month'].replace(4, 'Q2')
X['Month'] = X['Month'].replace(5, 'Q2')
X['Month'] = X['Month'].replace(6, 'Q2')
X['Month'] = X['Month'].replace(7, 'Q3')
X['Month'] = X['Month'].replace(8, 'Q3')
X['Month'] = X['Month'].replace(9, 'Q3')
X['Month'] = X['Month'].replace(10, 'Q4')
X['Month'] = X['Month'].replace(11, 'Q4')
X['Month'] = X['Month'].replace(12, 'Q4')

In [41]:
X = pd.get_dummies(X)

In [42]:
X.head()

AssessedValue  SalesRatio  Ansonia  Avon  Berlin  Bethel  Bloomfield  \
0         795870    0.866976        0     0       0       0           0   
1        1925560    1.674318        0     0       0       0           0   
2         189630    1.629489        0     0       0       0           0   
3         147340    1.018053        0     0       0       0           0   
4         163380    1.101726        0     0       0       0           0   

   Branford  Bridgeport  Bristol  ...  Woodstock  Four Family  Multi Family  \
0         0           0        0  ...          0            0             0   
1         0           0        0  ...          0            0             0   
2         0           0        0  ...          0            0             0   
3         0           0        0  ...          0            0             0   
4         0           1        0  ...          0            0             0   

   Single Family  Three Family  Two Family  Month_Q1  Month_Q2  Month_Q3  \
0              1             0           0         0         1         0   
1              1             0           0         0         0         0   
2              1             0           0         0         0         0   
3              1             0           0         0         1         0   
4              1             0           0         1         0         0   

   Month_Q4  
0         0  
1         1  
2         1  
3         0  
4         0  

[5 rows x 136 columns]

In [43]:
scaler.fit(X)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [44]:
X_scaled = scaler.transform(X)

In [45]:
X_scaled = sm.add_constant(X_scaled)

In [46]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size = 0.3, random_state = 123)

In [47]:
X_train.shape

(402300, 137)

In [48]:
X_test.shape

(172415, 137)

In [49]:
Model2 = sm.OLS(y_train.astype(float), X_train.astype(float))
Model2_results = Model2.fit()

In [50]:
Model2_results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             SaleAmount   R-squared:                       0.769
Model:                            OLS   Adj. R-squared:                  0.769
Method:                 Least Squares   F-statistic:                 1.007e+04
Date:                Wed, 24 Mar 2021   Prob (F-statistic):               0.00
Time:                        18:29:28   Log-Likelihood:            -5.4635e+06
No. Observations:              402300   AIC:                         1.093e+07
Df Residuals:                  402166   BIC:                         1.093e+07
Df Model:                         133                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       3.356e+05    301.694   1112.475      0.000    3.35e+05    3.36e+05
x1          2.824e+05    406.584    694.524      0.000    2.82e+05    2.83e+05
x2          1396.3083    297.393      4.695      0.000     813.427    1979.189
x3         -3574.2509    302.115    -11.831      0.000   -4166.388   -2982.114
x4          3826.1468    299.304     12.783      0.000    3239.519    4412.775
x5         -2540.7556    299.876     -8.473      0.000   -3128.504   -1953.008
x6          -610.4970    298.524     -2.045      0.041   -1195.596     -25.399
x7         -2002.7499    300.275     -6.670      0.000   -2591.279   -1414.221
x8         -1418.1587    298.771     -4.747      0.000   -2003.741    -832.576
x9         -1.233e+04    299.368    -41.181      0.000   -1.29e+04   -1.17e+04
x10        -6783.0180    297.370    -22.810      0.000   -7365.854   -6200.182
x11          837.5964    299.108      2.800      0.005     251.353    1423.840
x12        -1769.3938    302.970     -5.840      0.000   -2363.205   -1175.583
x13         -230.0802    298.544     -0.771      0.441    -815.217     355.057
x14          446.7842    303.659      1.471      0.141    -148.379    1041.947
x15         -106.7222    298.463     -0.358      0.721    -691.701     478.257
x16         -584.3362    299.188     -1.953      0.051   -1170.736       2.064
x17        -1859.9535    301.140     -6.176      0.000   -2450.180   -1269.727
x18        -1217.2185    295.846     -4.114      0.000   -1797.068    -637.369
x19        -3004.0848    298.466    -10.065      0.000   -3589.068   -2419.101
x20        -1803.1958    300.044     -6.010      0.000   -2391.273   -1215.118
x21         -265.5325    297.622     -0.892      0.372    -848.864     317.799
x22         3.498e+04    313.208    111.690      0.000    3.44e+04    3.56e+04
x23        -2564.5285    300.789     -8.526      0.000   -3154.065   -1974.992
x24         -626.3869    298.535     -2.098      0.036   -1211.507     -41.267
x25         -983.4010    305.168     -3.222      0.001   -1581.521    -385.281
x26        -1021.4806    299.597     -3.410      0.001   -1608.682    -434.280
x27        -1482.5648    302.220     -4.906      0.000   -2074.907    -890.222
x28        -8826.7994    298.527    -29.568      0.000   -9411.903   -8241.696
x29        -4391.1968    300.287    -14.623      0.000   -4979.750   -3802.644
x30         -501.2528    300.836     -1.666      0.096   -1090.882      88.377
x31        -2576.7271    295.226     -8.728      0.000   -3155.362   -1998.092
x32         3573.8766    306.444     11.662      0.000    2973.255    4174.498
x33         -731.7381    301.840     -2.424      0.015   -1323.336    -140.140
x34        -6064.8089    300.371    -20.191      0.000   -6653.528   -5476.090
x35         2194.9306    297.002      7.390      0.000    1612.817    2777.045
x3

In [51]:
y_pred = Model2_results.predict(X_test)

In [52]:
matches = pd.DataFrame(y_test)
matches.rename(columns = {'SaleAmount':'actual'}, inplace=True)
matches["predicted"] = y_pred

rmse(matches["actual"], matches["predicted"])

194199.838772732

Next we will run the model eliminating columns with a p-value that is above 0.05 these columns are: 13, 14, 15, 21, 30, 58, 70, 81, 102, 126, 129 (kept values that were still very close to 0.05)

In [53]:
X.drop(X.columns[[12, 13, 14, 20, 29, 57, 69, 80, 101, 125, 128]], axis = 1, inplace = True)

In [54]:
scaler.fit(X)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [55]:
X_scaled = scaler.transform(X)

In [56]:
X_scaled = sm.add_constant(X_scaled)

In [57]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size = 0.3, random_state = 123)

In [58]:
X_train.shape

(402300, 126)

In [59]:
X_test.shape

(172415, 126)

In [60]:
Model2 = sm.OLS(y_train.astype(float), X_train.astype(float))
Model2_results = Model2.fit()

In [61]:
Model2_results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             SaleAmount   R-squared:                       0.769
Model:                            OLS   Adj. R-squared:                  0.769
Method:                 Least Squares   F-statistic:                 1.081e+04
Date:                Wed, 24 Mar 2021   Prob (F-statistic):               0.00
Time:                        18:29:46   Log-Likelihood:            -5.4635e+06
No. Observations:              402300   AIC:                         1.093e+07
Df Residuals:                  402175   BIC:                         1.093e+07
Df Model:                         124                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       3.356e+05    301.692   1112.479      0.000    3.35e+05    3.36e+05
x1          2.824e+05    406.523    694.638      0.000    2.82e+05    2.83e+05
x2          1396.2923    297.392      4.695      0.000     813.413    1979.172
x3         -3365.4872    315.650    -10.662      0.000   -3984.152   -2746.823
x4          4073.5838    318.070     12.807      0.000    3450.176    4696.991
x5         -2326.6924    313.922     -7.412      0.000   -2941.969   -1711.416
x6          -398.2621    312.483     -1.275      0.202   -1010.720     214.196
x7         -1766.7274    317.201     -5.570      0.000   -2388.432   -1145.023
x8         -1190.9161    314.684     -3.784      0.000   -1807.686    -574.146
x9         -1.178e+04    384.111    -30.671      0.000   -1.25e+04    -1.1e+04
x10        -6402.5350    340.370    -18.810      0.000   -7069.651   -5735.419
x11         1049.7698    313.004      3.354      0.001     436.292    1663.248
x12        -1618.2871    309.982     -5.221      0.000   -2225.842   -1010.732
x13         -370.4052    313.234     -1.183      0.237    -984.335     243.524
x14        -1662.7899    313.067     -5.311      0.000   -2276.392   -1049.187
x15        -1091.8143    300.779     -3.630      0.000   -1681.332    -502.296
x16        -2803.2690    310.871     -9.017      0.000   -3412.567   -2193.971
x17        -1622.8284    310.071     -5.234      0.000   -2230.559   -1015.098
x18         3.525e+04    334.979    105.235      0.000    3.46e+04    3.59e+04
x19        -2403.6712    308.904     -7.781      0.000   -3009.114   -1798.228
x20         -489.8530    304.351     -1.609      0.108   -1086.372     106.666
x21         -857.0324    310.063     -2.764      0.006   -1464.747    -249.318
x22         -848.6716    308.820     -2.748      0.006   -1453.950    -243.393
x23        -1269.1700    316.091     -4.015      0.000   -1888.700    -649.640
x24        -8455.3260    339.369    -24.915      0.000   -9120.480   -7790.172
x25        -4115.2674    323.256    -12.731      0.000   -4748.839   -3481.695
x26        -2411.9833    303.719     -7.942      0.000   -3007.263   -1816.704
x27         3717.0534    312.792     11.883      0.000    3103.990    4330.117
x28         -533.3116    313.899     -1.699      0.089   -1148.544      81.920
x29        -5749.1413    330.147    -17.414      0.000   -6396.219   -5102.063
x30         2348.0257    304.365      7.715      0.000    1751.479    2944.572
x31          1.17e+04    346.747     33.744      0.000     1.1e+04    1.24e+04
x32         2505.1663    321.493      7.792      0.000    1875.051    3135.282
x33         2170.6376    326.419      6.650      0.000    1530.865    2810.410
x34         -378.1951    306.312     -1.235      0.217    -978.557     222.166
x35         5.046e+04    374.163    134.874      0.000    4.97e+04    5.12e+04
x3

I decided to continue eliminating columns with high p-values until all remaining were below 0.05

In [62]:
X.drop(X.columns[[5, 12, 19, 27, 33, 43, 50, 70, 87, 105, 117, 118, 119, 120]], axis = 1, inplace = True)

In [63]:
scaler.fit(X)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [64]:
X_scaled = scaler.transform(X)

In [65]:
X_scaled = sm.add_constant(X_scaled)

In [66]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size = 0.3, random_state = 123)

In [67]:
X_train.shape

(402300, 112)

In [68]:
X_test.shape

(172415, 112)

In [69]:
Model2 = sm.OLS(y_train.astype(float), X_train.astype(float))
Model2_results = Model2.fit()

In [70]:
Model2_results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             SaleAmount   R-squared:                       0.769
Model:                            OLS   Adj. R-squared:                  0.769
Method:                 Least Squares   F-statistic:                 1.218e+04
Date:                Wed, 24 Mar 2021   Prob (F-statistic):               0.00
Time:                        18:30:03   Log-Likelihood:            -5.4635e+06
No. Observations:              402300   AIC:                         1.093e+07
Df Residuals:                  402189   BIC:                         1.093e+07
Df Model:                         110                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       3.356e+05    301.726   1112.359      0.000    3.35e+05    3.36e+05
x1          2.824e+05    406.172    695.247      0.000    2.82e+05    2.83e+05
x2          1396.8906    297.427      4.697      0.000     813.944    1979.838
x3         -3242.3011    310.268    -10.450      0.000   -3850.418   -2634.184
x4          4323.1073    311.137     13.895      0.000    3713.288    4932.927
x5         -2118.4211    308.714     -6.862      0.000   -2723.492   -1513.350
x6         -1529.6747    310.917     -4.920      0.000   -2139.062    -920.287
x7          -979.4077    308.812     -3.172      0.002   -1584.670    -374.145
x8          -1.14e+04    349.292    -32.627      0.000   -1.21e+04   -1.07e+04
x9         -6053.7713    324.670    -18.646      0.000   -6690.115   -5417.428
x10         1262.4359    307.852      4.101      0.000     659.055    1865.816
x11        -1476.0406    307.387     -4.802      0.000   -2078.510    -873.571
x12        -1477.9978    308.656     -4.789      0.000   -2082.954    -873.042
x13         -967.5464    298.946     -3.237      0.001   -1553.471    -381.621
x14        -2603.4736    306.253     -8.501      0.000   -3203.720   -2003.228
x15        -1448.4052    306.349     -4.728      0.000   -2048.840    -847.971
x16         3.552e+04    327.043    108.619      0.000    3.49e+04    3.62e+04
x17        -2301.3463    305.742     -7.527      0.000   -2900.591   -1702.102
x18         -733.7592    308.258     -2.380      0.017   -1337.935    -129.584
x19         -678.9769    305.393     -2.223      0.026   -1277.539     -80.415
x20        -1064.8259    310.960     -3.424      0.001   -1674.297    -455.354
x21        -8137.6508    324.540    -25.074      0.000   -8773.740   -7501.562
x22        -3856.5136    314.809    -12.250      0.000   -4473.530   -3239.497
x23        -2256.2503    300.561     -7.507      0.000   -2845.341   -1667.160
x24         3856.3319    310.466     12.421      0.000    3247.829    4464.835
x25        -5466.8813    319.271    -17.123      0.000   -6092.643   -4841.120
x26         2499.5599    301.634      8.287      0.000    1908.366    3090.754
x27         1.209e+04    329.660     36.670      0.000    1.14e+04    1.27e+04
x28         2749.3160    314.625      8.738      0.000    2132.660    3365.972
x29         2475.0933    315.377      7.848      0.000    1856.965    3093.222
x30          5.08e+04    360.609    140.873      0.000    5.01e+04    5.15e+04
x31        -2747.5105    306.424     -8.966      0.000   -3348.093   -2146.928
x32        -2005.7096    314.413     -6.379      0.000   -2621.949   -1389.470
x33         2144.7935    314.614      6.817      0.000    1528.159    2761.428
x34        -1237.0766    307.052     -4.029      0.000   -1838.888    -635.265
x35        -4536.6102    325.994    -13.916      0.000   -5175.549   -3897.671
x3

In [71]:
X.drop(X.columns[[36]], axis = 1, inplace = True)

In [72]:
scaler.fit(X)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [73]:
X_scaled = scaler.transform(X)

In [74]:
X_scaled = sm.add_constant(X_scaled)

In [75]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size = 0.3, random_state = 123)

In [76]:
X_train.shape

(402300, 111)

In [77]:
X_test.shape

(172415, 111)

In [78]:
Model2 = sm.OLS(y_train.astype(float), X_train.astype(float))
Model2_results = Model2.fit()

In [79]:
Model2_results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             SaleAmount   R-squared:                       0.769
Model:                            OLS   Adj. R-squared:                  0.769
Method:                 Least Squares   F-statistic:                 1.229e+04
Date:                Wed, 24 Mar 2021   Prob (F-statistic):               0.00
Time:                        18:30:20   Log-Likelihood:            -5.4635e+06
No. Observations:              402300   AIC:                         1.093e+07
Df Residuals:                  402190   BIC:                         1.093e+07
Df Model:                         109                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       3.356e+05    301.727   1112.358      0.000    3.35e+05    3.36e+05
x1          2.824e+05    406.161    695.277      0.000    2.82e+05    2.83e+05
x2          1396.9512    297.427      4.697      0.000     814.003    1979.899
x3         -3225.4652    310.051    -10.403      0.000   -3833.156   -2617.774
x4          4342.7821    310.841     13.971      0.000    3733.543    4952.021
x5         -2101.2360    308.487     -6.811      0.000   -2705.861   -1496.611
x6         -1510.6849    310.641     -4.863      0.000   -2119.531    -901.839
x7          -961.2195    308.557     -3.115      0.002   -1565.982    -356.457
x8         -1.135e+04    347.962    -32.624      0.000    -1.2e+04   -1.07e+04
x9         -6023.0969    323.979    -18.591      0.000   -6658.086   -5388.108
x10         1279.3418    307.631      4.159      0.000     676.394    1882.289
x11        -1463.8770    307.273     -4.764      0.000   -2066.123    -861.631
x12        -1462.1441    308.462     -4.740      0.000   -2066.720    -857.568
x13         -957.4611    298.865     -3.204      0.001   -1543.228    -371.694
x14        -2587.3155    306.050     -8.454      0.000   -3187.164   -1987.467
x15        -1433.9148    306.186     -4.683      0.000   -2034.030    -833.800
x16         3.554e+04    326.733    108.785      0.000    3.49e+04    3.62e+04
x17        -2288.3796    305.611     -7.488      0.000   -2887.368   -1689.392
x18         -723.6164    308.178     -2.348      0.019   -1327.637    -119.596
x19         -665.0867    305.243     -2.179      0.029   -1263.354     -66.819
x20        -1047.6633    310.734     -3.372      0.001   -1656.693    -438.633
x21        -8107.6777    323.880    -25.033      0.000   -8742.474   -7472.882
x22        -3834.2948    314.436    -12.194      0.000   -4450.579   -3218.010
x23        -2242.9786    300.422     -7.466      0.000   -2831.796   -1654.161
x24         3867.6088    310.368     12.461      0.000    3259.296    4475.922
x25        -5441.4502    318.788    -17.069      0.000   -6066.266   -4816.635
x26         2511.7378    301.517      8.330      0.000    1920.773    3102.703
x27         1.212e+04    328.934     36.847      0.000    1.15e+04    1.28e+04
x28         2769.1153    314.329      8.810      0.000    2153.041    3385.190
x29         2500.3332    314.895      7.940      0.000    1883.148    3117.519
x30         5.083e+04    360.098    141.151      0.000    5.01e+04    5.15e+04
x31        -2733.0268    306.261     -8.924      0.000   -3333.290   -2132.764
x32        -1982.8972    314.018     -6.315      0.000   -2598.364   -1367.431
x33         2164.7709    314.312      6.887      0.000    1548.728    2780.813
x34        -1225.1252    306.941     -3.991      0.000   -1826.721    -623.530
x35        -4504.9989    325.264    -13.850      0.000   -5142.506   -3867.492
x3

In [80]:
matches = pd.DataFrame(y_test)
matches.rename(columns = {'SaleAmount':'actual'}, inplace=True)
matches["predicted"] = y_pred

rmse(matches["actual"], matches["predicted"])

194199.838772732

## Model 3

Next let's try a Random Forest model with less variables

In [81]:
X = df[['AssessedValue', 'SalesRatio']]
X.head()

AssessedValue  SalesRatio
0         795870    0.866976
1        1925560    1.674318
2         189630    1.629489
3         147340    1.018053
4         163380    1.101726

In [82]:
scaler.fit(X)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [83]:
X_scaled = scaler.transform(X)

In [84]:
from sklearn.ensemble import RandomForestRegressor

In [85]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size = 0.3, random_state = 123)

In [86]:
X_train.shape

(402300, 2)

In [87]:
X_test.shape

(172415, 2)

In [88]:
Model3 = RandomForestRegressor(random_state = 123)
Model3.fit(X_train, y_train)

C:\Users\bronc\anaconda3\lib\site-packages\ipykernel_launcher.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  


RandomForestRegressor(bootstrap=True, ccp_alpha=0.0, criterion='mse',
                      max_depth=None, max_features='auto', max_leaf_nodes=None,
                      max_samples=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      n_estimators=100, n_jobs=None, oob_score=False,
                      random_state=123, verbose=0, warm_start=False)

In [89]:
y_pred = Model3.predict(X_test)

In [90]:
r2_score(y_test, y_pred)

0.9996200059108645

In [91]:
matches = pd.DataFrame(y_test)
matches.rename(columns = {'SaleAmount':'actual'}, inplace=True)
matches["predicted"] = y_pred

rmse(matches["actual"], matches["predicted"])

7793.665618712889

The R-squared score and root mean squared error are both very good for this model. R-squared is very near 1 and
root mean squared error is by far the lowest we've seen so far

## Model 4

Finally let's do a model that focuses on regularization. I'll try this out with the same split we used for model 3

In [92]:
alpha_range = 10.**np.arange(-2,3)
alpha_range

array([1.e-02, 1.e-01, 1.e+00, 1.e+01, 1.e+02])

In [93]:
from sklearn.linear_model import RidgeCV
Model4 = RidgeCV(alphas = alpha_range, normalize = True, scoring = 'neg_mean_squared_error')

In [94]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size = 0.3, random_state = 123)

In [95]:
Model4.fit(X_train, y_train)

RidgeCV(alphas=array([1.e-02, 1.e-01, 1.e+00, 1.e+01, 1.e+02]), cv=None,
        fit_intercept=True, gcv_mode=None, normalize=True,
        scoring='neg_mean_squared_error', store_cv_values=False)

In [96]:
Model4.alpha_

0.01

In [97]:
y_pred = Model4.predict(X_test)

In [98]:
from sklearn import metrics

In [99]:
r2_score(y_test, y_pred)

0.729110543378975

In [100]:
np.sqrt(metrics.mean_squared_error(y_test, y_pred))

208089.13228230696

Here our R-squared is lower than our other models and our root mean squared error is also not any better than the ones before

## Conclusion

It looks like the Random Forest regressor was by far the best model. It had a much higher R-squared almost equalling one and the root mean squared error was the lowest by far. This would be the model to implement to predict sale amount.